### ISSUE TITLE
AmbiguousMatchException for properties declared with 'new' in GetPropertyEx

### ISSUE THREAD

**Issue by TimKras** [Catel/Catel#1325]
We did an update from an older version of Catel to the latest version an we encounter the following problem. The problem is related to this issue: [link](https://stackoverflow.com/questions/11443707/getproperty-reflection-results-in-ambiguous-match-found-on-new-property)

The issue is related to the Auditor functionality. `AuditingManager.Clear()` will temporarry solve the issus.
Please find an repro here: [CatelAuditTest.zip](https://github.com/Catel/Catel/files/3265665/CatelAuditTest.zip)

See the stackStrace:

```
   at System.RuntimeType.GetPropertyImpl(String name, BindingFlags bindingAttr, Binder binder, Type returnType, Type[] types, ParameterModifier[] modifiers)
   at System.Type.GetProperty(String name, BindingFlags bindingAttr)
   at Catel.Reflection.ReflectionExtensions.GetPropertyEx(Type type, String name, BindingFlags bindingFlags, Boolean allowExplicitInterfaceProperties)
   at Catel.Reflection.PropertyHelper.<>c__DisplayClass15_0.<GetPropertyInfo>b__0()
   at Catel.Caching.CacheStorage`2.<>c__DisplayClass30_0.<GetFromCacheOrFetch>b__0()
   at Catel.Caching.CacheStorage`2.ExecuteInLock[T](TKey key, Func`1 action)
   at Catel.Caching.CacheStorage`2.GetFromCacheOrFetch(TKey key, Func`1 code, ExpirationPolicy expirationPolicy, Boolean override)
   at Catel.Caching.CacheStorage`2.GetFromCacheOrFetch(TKey key, Func`1 code, Boolean override, TimeSpan expiration)
   at Catel.Reflection.PropertyHelper.GetPropertyInfo(Object obj, String property, Boolean ignoreCase)
   at Catel.Reflection.PropertyHelper.TryGetPropertyValue[TValue](Object obj, String property, Boolean ignoreCase, Boolean throwOnException, TValue& value)
   at Catel.Reflection.PropertyHelper.TryGetPropertyValue[TValue](Object obj, String property, Boolean ignoreCase, TValue& value)
   at Catel.Reflection.PropertyHelper.TryGetPropertyValue[TValue](Object obj, String property, TValue& value)
   at Catel.Reflection.PropertyHelper.TryGetPropertyValue(Object obj, String property, Object& value)
   at Catel.MVVM.Auditing.AuditingHelper.OnViewModelPropertyChanged(Object sender, PropertyChangedEventArgs e)
   at Catel.Data.ObservableObject.RaisePropertyChanged(Object sender, AdvancedPropertyChangedEventArgs e)
   at Catel.Data.ModelBase.RaisePropertyChanged(Object sender, PropertyChangedEventArgs e, Boolean updateIsDirty, Boolean isRefreshCallOnly)
...
```

Platform:
.NET version:

**Comment by GeertvanHorrik** [comment]
Please create a small repro, then we can add unit tests and actually fix this.

**Comment by TimKras** [comment]
Updated the inital comment.

**Comment by GeertvanHorrik** [comment]
Thanks, I didn't see the edit. Will investigate and update the ticket.

**Comment by GeertvanHorrik** [comment]
Seems to be happening inside ReflectionExtensions.type.cs::GetPropertyEx:

![image](https://user-images.githubusercontent.com/1246444/59101702-20986200-892a-11e9-9aa3-4dcb30dfd7f6.png)

Not sure what a good fix would be. A few options:

1. First call without FlattenHierarchy, then fall back to include FlattenHierarchy in case nothing is found, but I can see test cases where it would fail (cons: bad performance + invalid cases)
2. Do a call to GetPropertiesEx() and filter the first that is found (cons: bad performance)
3. Don't redeclare properties (cons: you probably need this code somewhere)
4. Check for ambiguousmatchexception, in that case fall back to 2 (pro: should work in all cases, cons: harder code to understand)

_Btw did you know you can enable SourceLink and step through the Catel code?_

**Comment by GeertvanHorrik** [comment]
Written unit test:

```
[TestCase]
public void PreventsAmbiguousMatchExceptionForAmbiguousProperties()
{
    // Note: see https://github.com/Catel/Catel/issues/1325

    var propertyInfo = typeof(TestViewModel).GetPropertyEx(nameof(TestViewModel.Variable1));

    Assert.IsNotNull(propertyInfo);
    Assert.AreEqual(typeof(IPerson), propertyInfo.PropertyType);
}
```

Now seeing whether a sensible fix can be written.

**Comment by GeertvanHorrik** [comment]
Fixed, please try Build - 5.11.0-alpha.90 via myget (ready in about 20 minutes).

**Comment by TimKras** [comment]
Thanks for the quick solution.
